File: 1971. Find if Path Exists in Graph.py
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = {}
        for edge in edges:
            if edge[0] not in graph:
                graph[edge[0]] = []
            if edge[1] not in graph:
                graph[edge[1]] = []
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])
        return self.dfs(graph, source, destination, set())
    def dfs(self, graph, curr, destination, visited):
        if curr == destination:
            return True
        visited.add(curr)
        for neighbour in graph[curr]:
            if neighbour not in visited and dfs(graph, neighbour, destination, visited):
                return True
        return False

File: 1971. Find if Path Exists in Graph.java
class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            graph.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
        }
        boolean[] visited = new boolean[n];
        return dfs(graph, source, destination, visited);
    }
    private boolean dfs(Map<Integer, List<Integer>> graph, int curr, int destination, boolean[] visited) {
        if (curr == destination) return true;
        visited[curr] = true;
        for (int next : graph.get(curr)) {
            if (visited[next]) continue;
            boolean answer = dfs(graph, next, destination, visited);
            if (answer) return true;
        }
        return false;
    }
}